graph-query-ir/src/main/java/oracle/pgql/lang/ir/QueryExpression.java

```diff
@@ -2081,6 +2082,89 @@ public void accept(QueryExpressionVisitor v) {
     }
   }
 
+  class SourceDestinationPredicate implements QueryExpression {
+
+    private VarRef vertexReference;
+
+    private VarRef edgeReference;
+
+    private boolean isSourcePredicate;
+
+    public SourceDestinationPredicate(VarRef vertexReference, VarRef edgeReference, boolean isSourcePredicate) {
+      this.vertexReference = vertexReference;
+      this.edgeReference = edgeReference;
+      this.isSourcePredicate = isSourcePredicate;
+    }
+
+    public VarRef getVertexReference() {
+      return vertexReference;
+    }
+
+    public void setVertexReference(VarRef vertexReference) {
+      this.vertexReference = vertexReference;
+    }
+
+    public VarRef getEdgeReference() {
+      return edgeReference;
+    }
+
+    public void setEdgeReference(VarRef edgeReference) {
+      this.edgeReference = edgeReference;
+    }
+
+    public boolean isSourcePredicate() {
+      return isSourcePredicate;
+    }
+
+    public void setSourcePredicate(boolean isSourcePredicate) {
+      this.isSourcePredicate = isSourcePredicate;
+    }
+
+    @Override
+    public String toString() {
+      return vertexReference + " IS " + (isSourcePredicate ? "SOURCE" : "DESTINATION") + " " + edgeReference;
```

Comment:
Is the "OF" keyword missing?

Reply:
right.. accidentally left out.

it still ended up working because the OF keyword is optional in PGQL because I believe an early SQL/PGQ draft version had it optional, but the final version has it mandatory. So I did not document the optionality when we published the PGQL 2.0 spec. I'm now also getting rid of the optionality (see last commit) for PGQL. Technically a breaking change but should be OK since it was never part of the spec and it is a new and still unused feature.